examples/nidigital_spi/measurement.py

```diff
@@ -0,0 +1,184 @@
+measurement_service = nims.MeasurementService(
+    service_config_path=service_directory / "NIDigitalSPI.serviceconfig",
+    version="0.1.0.0",
+    ui_file_paths=[service_directory / "NIDigitalSPI.measui"],
+)
+service_options = ServiceOptions()
+
+
+@measurement_service.register_measurement
+@measurement_service.configuration(
+    "pin_names",
+    nims.DataType.PinArray1D,
+    ["SPI_PINS"],
+    instrument_type=nims.session_management.INSTRUMENT_TYPE_NI_DIGITAL_PATTERN,
+)
+@measurement_service.configuration(
+    "specification_file_path", nims.DataType.Path, "Specifications.specs"
+)
+@measurement_service.configuration("levels_file_path", nims.DataType.Path, "PinLevels.digilevels")
+@measurement_service.configuration("timing_file_path", nims.DataType.Path, "Timing.digitiming")
+@measurement_service.configuration("pattern_file_path", nims.DataType.Path, "Pattern.digipat")
+# TODO: MeasurementLink UI Editor doesn't support arrays of Booleans
+@measurement_service.output("passing_sites", nims.DataType.Int32Array1D)
+@measurement_service.output("failing_sites", nims.DataType.Int32Array1D)
+def measure(
+    pin_names: Iterable[str],
+    specifications_file_path: str,
+    levels_file_path: str,
+    timing_file_path: str,
+    pattern_file_path: str,
+) -> Tuple:
+    """Test a SPI device using an NI Digital Pattern instrument."""
+    logging.info("Starting test: pin_names=%s", pin_names)
+
+    session_management_client = nims.session_management.Client(
+        grpc_channel=measurement_service.get_channel(
+            provided_interface=nims.session_management.GRPC_SERVICE_INTERFACE_NAME,
+            service_class=nims.session_management.GRPC_SERVICE_CLASS,
+        )
+    )
+
+    with contextlib.ExitStack() as stack:
+        reservation = stack.enter_context(
+            session_management_client.reserve_sessions(
+                context=measurement_service.context.pin_map_context,
+                pin_or_relay_names=pin_names,
+                instrument_type_id=nims.session_management.INSTRUMENT_TYPE_NI_DIGITAL_PATTERN,
+                # If another measurement is using the session, wait for it to complete.
+                # Specify a timeout to aid in debugging missed unreserve calls.
+                # Long measurements may require a longer timeout.
+                timeout=60,
+            )
+        )
+
+        if len(reservation.session_info) != 1:
+            measurement_service.context.abort(
+                grpc.StatusCode.INVALID_ARGUMENT,
+                f"Unsupported number of sessions: {len(reservation.session_info)}",
+            )
+
+        session_info = reservation.session_info[0]
+        session = stack.enter_context(_create_nidigital_session(session_info))
+
+        pin_map_context = measurement_service.context.pin_map_context
+        selected_sites_string = ",".join(f"site{i}" for i in pin_map_context.sites)
+        selected_sites = session.sites[selected_sites_string]
```

Comment:
In InstrumentStudio and MeasurementLink UI Editor, selecting "all sites" works great: it makes one call with sites=[0,1,2,3].

In TestStand, the site number is based on the test socket, so you have to enable the batch process model to do multi-site, and then it runs 4 separate measurements, one site at a time.

Cc: @rfriedma @jonathanmendez

Reply:
This was an FYI, so I'm marking it resolved.